From c14cbb6314275badc15560656175c30d79bda9e3 Mon Sep 17 00:00:00 2001
From: Yang Shiyu <yangshiyu9@huawei.com>
Date: Wed, 30 Nov 2022 16:00:50 +0800
Subject: [PATCH 0876/1398] [Huawei] ocfs2: add support for ADL(Atomic Disk
 Lock) feature: add CDB

Offering: HULK
ocfs2 inclusion
category: feature
bugzilla: NA

----------------------------------------

Add CDB module for adl to send SCSI command to block level.

Signed-off-by: Yang Shiyu <yangshiyu9@huawei.com>
---
 fs/ocfs2/adl/cdb.c | 126 +++++++++++++++++++++++++++++++++++++++++++++
 fs/ocfs2/adl/cdb.h |  21 ++++++++
 2 files changed, 147 insertions(+)
 create mode 100644 fs/ocfs2/adl/cdb.c
 create mode 100644 fs/ocfs2/adl/cdb.h

diff --git a/fs/ocfs2/adl/cdb.c b/fs/ocfs2/adl/cdb.c
new file mode 100644
index 000000000000..3dff272ed383
--- /dev/null
+++ b/fs/ocfs2/adl/cdb.c
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include <scsi/scsi.h>
+#include <scsi/scsi_cmnd.h>
+#include <asm/unaligned.h>
+
+#include "cdb.h"
+
+#define MLOG_MASK_PREFIX ML_ADL
+#include "cluster/masklog.h"
+
+#define COMPARE_AND_WRITE_OPCODE 0x89
+
+struct cdb_request {
+	int	error;
+	struct completion waiting;
+};
+
+void setup_compare_and_write_cmd(struct caw_cdb *caw,
+		sector_t sector, unsigned int nr_sectors)
+{
+	memset(caw, 0x0, sizeof(struct caw_cdb));
+	caw->opc = COMPARE_AND_WRITE_OPCODE;
+	put_unaligned_be64(sector, caw->lba);
+	caw->length = (unsigned char)(nr_sectors & 0xff);
+}
+
+/* return -EAGAIN if miscompare with the data on disk */
+static int cdb_check_sense(struct request *rq)
+{
+	char *sense_buffer = scsi_req(rq)->sense;
+	int ret = -EIO;
+
+	if (!sense_buffer)
+		goto bail;
+
+	if (status_byte(scsi_req(rq)->result) == CHECK_CONDITION) {
+		u8 response_code = (sense_buffer[0] & 0x7f);
+		u8 sense_key;
+
+		/* sense invalid, return EIO */
+		if ((response_code & 0x70) != 0x70) {
+			ret = -EIO;
+			goto bail;
+		}
+		if (response_code >= 0x72)
+			sense_key = (sense_buffer[1] & 0xf);
+		else
+			sense_key = (sense_buffer[2] & 0xf);
+
+		if (sense_key == MISCOMPARE)
+			ret = -EAGAIN;
+	}
+
+bail:
+	return ret;
+}
+
+static void cdb_end_rq(struct request *rq, blk_status_t error)
+{
+	struct cdb_request *cdb_rq = rq->end_io_data;
+	int result = scsi_req(rq)->result;
+
+	if (result) {
+		cdb_rq->error = cdb_check_sense(rq);
+		if (cdb_rq->error != -EAGAIN) {
+			mlog(ML_ERROR,
+				"scsi_req(rq)->result 0x%x, status_byte 0x%x, host_byte 0x%x, driver_byte 0x%x\n",
+				result, status_byte(result),
+				host_byte(result), driver_byte(result));
+		}
+	} else if (error) {
+		mlog_errno(blk_status_to_errno(error));
+		cdb_rq->error = -EIO;
+	}
+
+	complete(&cdb_rq->waiting);
+	rq->end_io_data = NULL;
+#ifdef ARCH_TYPE_IS_X86
+	blk_put_request(rq);
+#endif  /* ARCH_TYPE_IS_X86 */
+#ifdef ARCH_TYPE_IS_ARM
+	__blk_put_request(rq->q, rq);
+#endif  /* ARCH_TYPE_IS_ARM */
+}
+
+int cdb_exec(struct block_device *bdev, int rw,
+		char *buf, unsigned int buf_len,
+		unsigned char *cmd)
+{
+	int ret;
+	struct cdb_request cdb_rq;
+	struct request *rq;
+	struct request_queue *q = bdev->bd_disk->queue;
+
+	if (buf_len > (queue_max_hw_sectors(q) << 9) || !buf_len)
+		return -EINVAL;
+
+	rq = blk_get_request(q, rw, 0);
+	if (IS_ERR(rq)) {
+		ret = PTR_ERR(rq);
+		mlog_errno(ret);
+		return ret;
+	}
+
+	cdb_rq.error = 0;
+	init_completion(&cdb_rq.waiting);
+
+	scsi_req(rq)->cmd_len = COMMAND_SIZE(cmd[0]);
+	memcpy(scsi_req(rq)->cmd, cmd, scsi_req(rq)->cmd_len);
+	rq->__sector = 0;
+	rq->timeout = START_STOP_TIMEOUT;
+	scsi_req(rq)->retries = 0;
+
+	ret = blk_rq_map_kern(q, rq, buf, buf_len, GFP_NOFS);
+	if (ret) {
+		mlog_errno(ret);
+		blk_put_request(rq);
+		return ret;
+	}
+
+	rq->end_io_data = &cdb_rq;
+	blk_execute_rq_nowait(q, bdev->bd_disk, rq, 0, cdb_end_rq);
+	wait_for_completion(&cdb_rq.waiting);
+
+	return cdb_rq.error;
+}
diff --git a/fs/ocfs2/adl/cdb.h b/fs/ocfs2/adl/cdb.h
new file mode 100644
index 000000000000..3ec5a3320f98
--- /dev/null
+++ b/fs/ocfs2/adl/cdb.h
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+#ifndef _CDB_H
+#define _CDB_H
+
+struct caw_cdb {
+	u8 opc;
+	u8 mask;
+	u8 lba[8];
+	u8 reserved1[3];
+	u8 length;
+	u8 reserved2;
+	u8 control;
+};
+
+void setup_compare_and_write_cmd(struct caw_cdb *caw,
+		sector_t sector, unsigned int nr_sectors);
+
+int cdb_exec(struct block_device *bdev, int rw,
+		char *buf, unsigned int buf_len,
+		unsigned char *cmd);
+#endif
-- 
2.17.1

